src/date.c

```diff
@@ -1,18 +1,1006 @@
+  
+  ecma_getfirstdays(days, ily);
+  
+  for(i = 0; i < ARRAY_SIZE(days) - 1; i++) {
+    if(dwy >= days[i] && dwy < days[i+1]) {
+      ret = i;
+      break;
+    }
+  }
+  
+  return ret;
+}
+
+static int ecma_DateFromTime(etime_t t, int year) {
+  int days[MonthInYear+1];
+  int mft = ecma_MonthFromTime(t, year);
+  int dwy = ecma_DayWithinYear(t, year);
+  int ily = ecma_InLeapYear(t, year);
+  
+  if(mft > 11) {
+    return -1;
+  }
+  
+  ecma_getfirstdays(days, ily);
+  
+  return dwy - days[mft] + 1;
+}
+
+static int ecma_WeekDay(etime_t t) {
+  return (ecma_Day(t)+4) % 7;
+}
+
+static int ecma_DaylightSavingTA(etime_t t) {
+  time_t time = t / 1000;
+  struct tm tm;
+  memset(&tm, 0, sizeof(t));
+  localtime_r(&time, &tm);
+  if(tm.tm_isdst > 0) {
+    return msPerHour;
+  } else {
+    return 0;
+  }
+}
+
+static int ecma_LocalTZA() {
+  return (int)-timezone * 1000;
+}
+
+static etimeint_t ecma_LocalTime(etime_t t) {
+  return t + ecma_LocalTZA() + ecma_DaylightSavingTA(t);
+}
+
+static etimeint_t ecma_UTC(etime_t t) {
+  return t - ecma_LocalTZA() - ecma_DaylightSavingTA(t-ecma_LocalTZA());
+}
+
+static int ecma_HourFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerHour) % HoursPerDay;
+}
+
+static int ecma_MinFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerMinute) % MinutesPerHour;
+}
+
+static int ecma_SecFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerSecond) % SecondsPerMinute;
+}
+
+static int ecma_msFromTime(etime_t t) {
+  return (etimeint_t)t % msPerSecond;
+}
+
+static etimeint_t ecma_MakeTime(etimeint_t hour, etimeint_t min,
+                                    etimeint_t sec, etimeint_t ms) {
+  return ((hour * MinutesPerHour + min)
+                      * SecondsPerMinute + sec) * msPerSecond + ms;
+}
+
+static etimeint_t ecma_MakeDay(int year, int month, int date)
+{
+  int days[MonthInYear+1];
+
+  etimeint_t yday;
+  etimeint_t mday;
+  
+  year += floor(month / 12);
+  
+  month = month % 12;
+  
+  yday = floor(ecma_TimeFromYear(year) / msPerDay);
+
+  ecma_getfirstdays(days, (ecma_DaysInYear(year) == 366));
+
+  mday = days[month];
+  
+  return yday + mday + date - 1;
+}
+
+static etimeint_t ecma_MakeDate(etimeint_t day, etimeint_t time) {
+  return (day * msPerDay + time);
+}
+
+/*++++ ECMA alternatives to struct tm ++++*/
+
+struct timeparts {
+  int year; /* can be negative, up to +-282000 */
+  int month; /* 0-11 */
+  int day; /* 1-31 */
+  int hour; /* 0-23 */
+  int min; /* 0-59 */
+  int sec; /* 0-59 */
+  int msec; 
+  int dayofweek; /* 0-6 */
+};
+
+static int d_istimeinvalid(etime_t* time) {
+  return isnan(*time);
+}
+
+/*+++ this functions is used to get current date/time & timezone */
+
+static void d_gettime(etime_t* time) {
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  *time = (etime_t)tv.tv_sec * 1000 + (etime_t)tv.tv_usec / 1000;
+}
+
+static int d_gettimezone() {
+  return (int)timezone/60; /* TODO(alashkin): check non-OSX */
```

Comment:
Our dev machines are OSX as well. We found a simple way of testing the code on linux:

Install boot2docker and run `make docker`. It will build and run all unit tests from right from your checkout.